test/common/TestCommonEqEpsilon.hs

```diff
@@ -70,26 +198,15 @@ assertCloseElem preface expected actual = do
       if abs (h-actual) < fromRational eqEps then assertClose msg h actual else go_assert eqEps t
 
 -- | Asserts that the specified actual floating point value list is close to the expected value.
-assertCloseList :: forall a. (AssertClose a, HasCallStack)
+assertCloseList :: forall a. (AssertClose a)
                 => [a]      -- ^ The expected value
                 -> [a]      -- ^ The actual value
                 -> Assertion
-assertCloseList expected actual =
-  go_assert expected actual
-  where
-    len1 :: Int = length expected
-    len2 :: Int = length actual
-    msgneq :: String = "expected " ++ show len1 ++ " elements, but got " ++ show len2
-    go_assert :: [a] -> [a] -> Assertion
-    go_assert [] [] = assertBool "" True
-    go_assert [] (_:_) = assertFailure msgneq
-    go_assert (_:_) [] = assertFailure msgneq
-    go_assert (head_exp:tail_exp) (head_act:tail_act) =
-      (@?~) head_act head_exp >> go_assert tail_exp tail_act
+assertCloseList = assert_list (flip (@?~))
 
--- | Foldable to list.
-asList :: Foldable t => t a -> [a]
-asList = foldr (:) []
+----------------------------------------------------------------------------
+-- AssertClose class together with (@?~) operator
```

Comment:
Why not define `@?~` in terms of `assertEqualUpToEpsilon`?

Reply:
I am trying things similar to the snippet below. But apparently I am having problems convincing Haskell to unify the type variables `z` and `z0`:

```haskell
instance {-# OVERLAPPABLE #-} forall a z . (Fractional z, AssertEqualUpToEpsilon z a) => AssertClose a where
  (@?~) :: forall {z} . a -> a -> Assertion
  (@?~) actual expected = do
    eqEpsilon <- readIORef eqEpsilonRef :: Rational
    assertEqualUpToEpsilon ((fromRational (eqEpsilon :: Rational)) :: z) (expected :: a) (actual :: a)
```

This is pretty verbose with respect to types, as I was trying to disambiguate it as much as I could. But still I am getting this error:

```
test/common/TestCommonEqEpsilon.hs:223:31: error:
    • Could not deduce (Fractional z0)
      from the context: (Fractional z, AssertEqualUpToEpsilon z a)
        bound by an instance declaration:
                   forall a z.
                   (Fractional z, AssertEqualUpToEpsilon z a) =>
                   AssertClose a
        at test/common/TestCommonEqEpsilon.hs:223:31-102
      The type variable ‘z0’ is ambiguous
    • In the ambiguity check for an instance declaration
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      In the instance declaration for ‘AssertClose a’
    |
223 | instance {-# OVERLAPPABLE #-} forall a z . (Fractional z, AssertEqualUpToEpsilon z a) => AssertClose a where
    |
```

What am I doing wrong here?

Reply:
First, why not remove the `AssertClose` class altogether and use only `AssertEqualUpToEpsilon`? Second, this is tough in general, but at this point I'd try adding the `Fractional` constraint to `(@?~)`. The variable `z` in this function has no relation to the outer `z` (because it's in an inner binding) and so doesn't have the constraint, so GHC correctly complains.

Reply:
Or perhaps remove the inner `forall {z}`, but I guess you tried that already. However, the error message may then be more to the point.